### PR TITLE
Feat/repo privatisation

### PR DESCRIPTION
**Pull request by alexanderleegs**
**NOTE**
This feature is currently disabled on the end user side while we work on blocking access to the netlify site of privatised sites - will be added in a later pr.

Video for end to end user flow:
https://github.com/isomerpages/isomercms-frontend/assets/22111124/97a1d624-1e8a-479e-81b5-eb7a0d8f292f

## Problem

This PR introduces a repo privatisation feature to the settings page for repos. To be reviewed in conjunction with PR #[806](https://github.com/isomerpages/isomercms-backend/pull/806) on the isomercms-backend repo.

The repo privatisation field should only show up for Amplify sites. This solution automatically generates a password when first enabled by the user, but can be manually changed by the user. 

## Tests

- [ ] Navigate to the settings page of an amplify site (which is recorded in your local db, with entries in `sites` and `deployments`
- [ ] Test password functionality
  - [ ] Can add password
    - [ ] Verify that password works on amplify site and github repo is set to private
  - [ ] Can change password
    - [ ] Verify that new password works on amplify site and github repo is still private
  - [ ] Can remove password
    - [ ] Verify that no password needed for amplify site and github repo is set to public

Before deployment 

- DONE - The IAM role associated to our ec2 instances needs to add additional permissions to be able to modify amplify apps - we will need to modify the prod role as well before merging

### CODE REVIEW

File: src/layouts/Settings/PrivacySettings.tsx
Comment (harishv7) [inline src/layouts/Settings/PrivacySettings.tsx:50]: would it be easier to just use a password generation lib ? https://www.npmjs.com/package/generate-password

File: src/layouts/Settings/PrivacySettings.tsx
Comment (harishv7) [inline src/layouts/Settings/PrivacySettings.tsx:76]: Shall we add a todo to replace this after we up design system?

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx]: Yup, that makes a lot of sense, have swapped it out in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/9463fa9a4c9d7791f8d8b41935cd03ef5f111a01

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx]: I've added the TODO, but for a slightly different change - after having a chat with Nat, we're opting to swap all the existing toggle buttons in settings to checkboxes instead! Have added a TODO for that instead to swap all at one time since that is out of scope of this PR

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts]: Changed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/a69a220cbfa2f40c0feaeba82c93ce9a2cb20aca

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts]: Have changed it to `Promise<void | null>` given the change to return null instead in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/a69a220cbfa2f40c0feaeba82c93ce9a2cb20aca

File: src/theme/foundations/colours.ts
Comment (alexanderleegs) [reply src/theme/foundations/colours.ts:12]: Hmm this is the colour variable name from figma, will double check with Nat that the colour is okay!

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:16]: We've actually been using camelCase for our routes currently (e.g. `/resourceRoom`) - for the sake of consistency, what do you feel about leaving it as is for now, and maybe swapping all of them to snake-case in a future PR?

**Review by kishore03109 — CHANGES_REQUESTED** [review]
Upon reading more into this topic, I am afraid I am concerned about the implementation of this feature :sadge: (ps, should have brought it up during the RFC discussion)

It seems that we shouldn't be storing [sensitive data in our frontend](https://stackoverflow.com/questions/48699820/how-do-i-hide-an-api-key-in-create-react-app). To quote the [docs](https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env): `WARNING: Do not store any secrets (such as private API keys) in your React app!` 

That means that in essence, we should only be decrypting and encrypting them on the server side, and send unencrypted password over the encrypted network. 

please let me know if you feel that my assessment is inaccurate!

File: .env-example
Comment (alexanderleegs) [reply .env-example:21]: As discussed offline - removing this env var!

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx:92]: No choice for this one - I don't think it makes sense to manually copy out all the methods from react-hook-form here!

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:71]: This has been removed after the change to encryption/decryption!

File: src/utils/password.ts
Comment (alexanderleegs) [reply src/utils/password.ts:49]: Removed, see above

File: src/utils/password.ts
Comment (alexanderleegs) [reply src/utils/password.ts:36]: Removed, see above

File: src/utils/password.ts
Comment (alexanderleegs) [reply src/utils/password.ts:41]: This is the usage as given by the [documentation](https://nodejs.org/docs/latest-v18.x/api/crypto.html#class-decipher) - i'm not sure what to call the intermediate steps either tbh, so I used their naming. But .update alone doesn't actually provide the final decrypted string, and .final is needed to cut off the decipher usage, but doesn't actually provide any decryption

File: src/utils/password.ts
Comment (alexanderleegs) [reply src/utils/password.ts:13]: Removed! This was originally used to generate the secret key, forgot to remove afterwards

File: src/utils/password.ts
Comment (alexanderleegs) [reply src/utils/password.ts]: As discussed offline, we're going to send passwords in plaintext and rely on transport layer security for now (since this is sent over https) - we'll eventually add extra redundancies and checks, but for the first cut we'll go with this

Modified in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/5487b17f699e291220087f30b3485344d71087fc!

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:62]: Most of this has been removed, moved the remaining guard clause up early in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/a30eb6cfd8b627be0b7e18aac2ef80f207489f12

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx]: This is no longer needed - was using it originally for a useeffect that triggered only on page load. Removed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/056c275869c433b7c0cff31d9794bad6c16cb1ec

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts]: Oh we already have a couple of kebab-case endpoints - in that case i've changed this to kebab-case in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/5eb9ae8df489aff541e906380ddb98a9d21a0213!

File: src/services/SettingsService.ts
Comment (harishv7) [inline src/services/SettingsService.ts:14]: hey alex i thought we were keeping the route camelCase? just wanted to check if we have changed both the route on FE & BE?

File: src/layouts/Settings/PrivacySettings.tsx
Comment (kishore03109) [reply src/layouts/Settings/PrivacySettings.tsx:91]: as in this comment was to add the ignore eslint rule thingi, but this has irked me enuf to create this [pr](https://github.com/isomerpages/isomercms-frontend/pull/1328) hahah

File: package.json
Comment (harishv7) [inline package.json:29]: do we still need these changes?

File: craco.config.js
Comment (harishv7) [inline craco.config.js:19]: can we remove this after shifting to window.crypto?

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:14]: I realised we already use some kebab-case endpoints (`/launch-site`), so it's already inconsistent! I swapped it on both FE and BE

File: src/types/settings.ts
Comment (alexanderleegs) [reply src/types/settings.ts]: Changed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/edfff02840b94ded1855fd4ca76d73ac5e0d8a35!

File: craco.config.js
Comment (alexanderleegs) [reply craco.config.js]: Yup, removed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/6307c6951eede47ad3e35791c308cf44246db347

File: package.json
Comment (alexanderleegs) [reply package.json]: Removed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/a8d740317006ec36f098d92030b3f66ec4c4b37e!

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (alexanderleegs) [reply src/hooks/settingsHooks/useGetSettings.ts:8]: Have filed an issue!

File: src/services/SettingsService.ts
Comment (harishv7) [reply src/services/SettingsService.ts:14]: got it thanks!

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx:105]: Ah right! Have removed it in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/46b7be69c05fc25285d1503fc96eea83f4c192bf

File: .env-example
Comment (kishore03109) [reply .env-example:21]: hey I think you might have forgotten to remove this

File: src/services/SettingsService.ts
Comment (kishore03109) [inline src/services/SettingsService.ts:50]: [ts error]: Parameter 'res' implicitly has an 'any' type.

File: src/services/SettingsService.ts
Comment (kishore03109) [inline src/services/SettingsService.ts:55]: is this placed here by accident?

File: src/types/settings.ts
Comment (kishore03109) [inline src/types/settings.ts:65]: why is this set to `boolean` ah? isnt this always true?

**Review by kishore03109 — COMMENTED** [review]
https://github.com/isomerpages/isomercms-backend/assets/42832651/f9b7009a-6dbf-45cf-9b01-66eb6c59c188

Not sure if I am doing this wrongly but this doesnt seem to work as expected in staging :( 


1. Save button does not seem to work + dont see any network calls 
2. the `link copied` tooltip stays there permanently
3. Resulting repos are not priviate + resulting amp repo is not private as well :( 

did we miss out a bug in staging?

File: .env-example
Comment (alexanderleegs) [reply .env-example:21]: ad3454b4

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:55]: Whoops yeah, removed in 7d2e0f78

File: src/types/settings.ts
Comment (alexanderleegs) [reply src/types/settings.ts:65]: This is a distinguishing boolean for whether a site is amplify or not! netlify sites will have this as `false`

File: src/services/SettingsService.ts
Comment (alexanderleegs) [reply src/services/SettingsService.ts:50]: I'm not getting this on my end - the type should be specified by `apiService
    .get<BackendPasswordSettings>`

Comment (kishore03109) [reply src/services/SettingsService.ts:50]: ya can ignore this sorry error in my local machine

File: src/types/settings.ts
Comment (kishore03109) [reply src/types/settings.ts:65]: wait I dont get this ya, isnt this why we should be setting 
```
interface BackendPasswordAmplifySettings {
  password: string
  isAmplifySite: true
}
```
because we know that all password sites are definitely amplify sites?

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (harishv7) [inline src/hooks/settingsHooks/useGetSettings.ts:137]: why can't we do convertFromBe(siteSettings)?

File: src/layouts/Settings/PrivacySettings.tsx
Comment (harishv7) [reply src/layouts/Settings/PrivacySettings.tsx:105]: i still see the remove the props-spreading disable line

File: src/theme/foundations/colours.ts
Comment (harishv7) [reply src/theme/foundations/colours.ts:34]: have we checked this?

File: src/types/settings.ts
Comment (harishv7) [reply src/types/settings.ts:65]: yea it should be a fixed `true` for amplify sites - essentially creating the discriminating union

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (alexanderleegs) [reply src/hooks/settingsHooks/useGetSettings.ts]: Oops we can now, changed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/fd7ce4c772e35f616905cf5b54e49f773857b8a4

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx:106]: hmm it should be gone after the change in #1331 !

File: src/theme/foundations/colours.ts
Comment (alexanderleegs) [reply src/theme/foundations/colours.ts]: Ah i realised we weren't actually using it - have removed it in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/9e9afeafd20bd8e2217bf5a93ab2264a25847d9a!

**Review by kishore03109 — COMMENTED** [review]
hey have we decided on a strategy for handling netlify sites? 
how will this affect the current repo privatisation PRs? are we still pushing them in first?

File: src/theme/foundations/colours.ts
Comment (alexanderleegs) [reply src/theme/foundations/colours.ts]: I think i used it in one of the commits, then removed it and forgot to remove the token!

File: package-lock.json
Comment (harishv7) [inline package-lock.json:1787]: @alexanderleegs since there are no package.json changes are the lock file changes needed/intended?

File: .storybook/main.js
Comment (harishv7) [inline .storybook/main.js:5]: can remove this now right?

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (harishv7) [inline src/hooks/settingsHooks/useGetSettings.ts:137]: when should we re-enable this?

File: src/layouts/Settings/PrivacySettings.tsx
Comment (harishv7) [reply src/layouts/Settings/PrivacySettings.tsx:106]: oh i meant we no longer need to add the ` // eslint-disable-next-line react/jsx-props-no-spreading` comment right?

File: package-lock.json
Comment (alexanderleegs) [reply package-lock.json]: yup, it should be fine, removed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/68ec476d487f67d7bce2ba4e9731b80686c9d980!

File: .storybook/main.js
Comment (alexanderleegs) [reply .storybook/main.js]: yup, https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/d7782a3614511cb981589a03759f25fbadcbbce3

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (alexanderleegs) [reply src/hooks/settingsHooks/useGetSettings.ts:137]: There's an associated backend PR up (https://github.com/isomerpages/isomercms-backend/pull/828) - once it's approved we can re-enable this!

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx:105]: Oh yeah, removed in https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/4e3c34b806a39ab6eee3e6dbe4d752156d996812

**Review by kishore03109 — APPROVED** [review]
mostly lgtm, but may I check in where is the pr to merge undo the netlify stuff? if still not up yet, please dismiss this review

File: src/layouts/Settings/PrivacySettings.tsx
Comment (harishv7) [inline src/layouts/Settings/PrivacySettings.tsx:99]: nit: shouldnt this have a fullstop similar to other labels?

**Review by harishv7 — APPROVED** [review]

File: src/hooks/settingsHooks/useGetSettings.ts
Comment (alexanderleegs) [reply src/hooks/settingsHooks/useGetSettings.ts:137]: Hey - after further discussion, we'll be keeping this feature hidden for now! The netlify change is in https://github.com/isomerpages/isomercms-backend/pull/828, the conditional has been changed in https://github.com/isomerpages/isomercms-frontend/pull/1344 (but is still disabled for all sites)

File: src/layouts/Settings/PrivacySettings.tsx
Comment (alexanderleegs) [reply src/layouts/Settings/PrivacySettings.tsx]: Oh it should, https://github.com/isomerpages/isomercms-frontend/pull/1316/commits/1f72a35279945dabf24db1903ee56ea778260075

**Pull request merged**
